### ISSUE TITLE
[Bug] lower mAP with higher batch after deploying a rotated object detection model

### ISSUE THREAD

**Issue by dimakan** [open-mmlab/mmdeploy#1945]
### Checklist

- [X] I have searched related issues but cannot get the expected help.
- [X] 2. I have read the [FAQ documentation](https://github.com/open-mmlab/mmdeploy/blob/master/docs/en/faq.md) but cannot get the expected help.
- [X] 3. The bug has not been fixed in the latest version.

### Describe the bug

After converting pretrained or custom rotated pth to tensorrt with dynamic batch, getting worse results with test.py and Triton Inference server.
Getting less detections and mAP drop with higher batch sizes.
Happens on 0.x and 1.x.

Tried to create a fixed shape with batch 8, same result as dynamic 8.

Tried to reproduce with sdk example by duplicating input (for simplicity), but result seems consistent (multiple of batch).

### Reproduction

**Deploy**
python $MMDEPLOY_DIR/tools/deploy.py \
    mmdeploy/configs/mmrotate/rotated-detection_tensorrt-fp16_dynamic-320x320-1024x1024-batch.py \
    $MMROTATE_DIR/configs/roi_trans/roi-trans-le90_r50_fpn_1x_dota.py \
    checkpoints/roi_trans_r50_fpn_1x_dota_le90-d1f0b77a.pth \
    $MMROTATE_DIR/demo/demo.jpg \
    --work-dir work-dirs/mmrotate/roi_trans/ort/dyn_size_batch \
    --dump-info \
    --device cuda

**Test b=1**
python $MMDEPLOY_DIR/tools/test.py \
    mmdeploy/configs/mmrotate/rotated-detection_tensorrt-fp16_dynamic-320x320-1024x1024-batch.py \
    $MMROTATE_DIR/configs/roi_trans/roi-trans-le90_r50_fpn_1x_dota.py \
    --model work-dirs/mmrotate/roi_trans/ort/dyn_size_batch/end2end.engine \
    --batch-size 1 \
    --speed-test \
    --device cuda
>> 03/30 22:01:04 - mmengine - INFO - Epoch(test) [448/448]  **dota/mAP: 0.3384**  dota/AP50: 0.3380data_time: 0.0366  time: 0.0528  

**Test b=4**
python $MMDEPLOY_DIR/tools/test.py \
    mmdeploy/configs/mmrotate/rotated-detection_tensorrt-fp16_dynamic-320x320-1024x1024-batch.py \
    $MMROTATE_DIR/configs/roi_trans/roi-trans-le90_r50_fpn_1x_dota.py \
    --model work-dirs/mmrotate/roi_trans/ort/dyn_size_batch/end2end.engine \
    --batch-size 4 \
    --speed-test \
    --device cuda

>> 03/30 21:26:38 - mmengine - INFO - Epoch(test) [112/112]  **dota/mAP: 0.1414**  dota/AP50: 0.1410data_time: 0.0895  time: 0.1370  

**Test b=8**
python $MMDEPLOY_DIR/tools/test.py \
    mmdeploy/configs/mmrotate/rotated-detection_tensorrt-fp16_dynamic-320x320-1024x1024-batch.py \
    $MMROTATE_DIR/configs/roi_trans/roi-trans-le90_r50_fpn_1x_dota.py \
    --model work-dirs/mmrotate/roi_trans/ort/dyn_size_batch/end2end.engine \
    --batch-size 8 \
    --speed-test \
    --device cuda

>> 03/30 22:03:47 - mmengine - INFO - Epoch(test) [56/56]  **dota/mAP: 0.1000**  dota/AP50: 0.1000data_time: 0.2529  time: 0.3423  

diff mmdeploy/configs/mmrotate/rotated-detection_tensorrt_dynamic-320x320-1024x1024.py mmdeploy/configs/mmrotate/rotated-detection_tensorrt_dynamic-320x320-1024x1024-batch.py

30,31c30,31
<                     opt_shape=[1, 3, 1024, 1024],
<                     max_shape=[1, 3, 1024, 1024])))
---
>                     opt_shape=[8, 3, 1024, 1024],
>                     max_shape=[16, 3, 1024, 1024])))

### Environment

```Shell
03/30 21:50:49 - mmengine - INFO - 

03/30 21:50:49 - mmengine - INFO - **********Environmental information**********
03/30 21:50:49 - mmengine - INFO - sys.platform: linux
03/30 21:50:49 - mmengine - INFO - Python: 3.8.16 (default, Mar  2 2023, 03:21:46) [GCC 11.2.0]
03/30 21:50:49 - mmengine - INFO - CUDA available: True
03/30 21:50:49 - mmengine - INFO - numpy_random_seed: 2147483648
03/30 21:50:49 - mmengine - INFO - GPU 0: NVIDIA GeForce RTX 3090
03/30 21:50:49 - mmengine - INFO - CUDA_HOME: /usr/local/cuda
03/30 21:50:49 - mmengine - INFO - NVCC: Cuda compilation tools, release 11.6, V11.6.124
03/30 21:50:49 - mmengine - INFO - GCC: gcc (Ubuntu 9.4.0-1ubuntu1~20.04.1) 9.4.0
03/30 21:50:49 - mmengine - INFO - PyTorch: 1.10.0
03/30 21:50:49 - mmengine - INFO - PyTorch compiling details: PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2020.0.2 Product Build 20200624 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v2.2.3 (Git Hash 7336ca9f055cf1bfa13efb658fe15dc9b41f0740)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - LAPACK is enabled (usually provided by MKL)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.3
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_37,code=compute_37
  - CuDNN 8.2
  - Magma 2.5.2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.3, CUDNN_VERSION=8.2.0, CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -DSYMBOLICATE_MOBILE_DEBUG_HANDLE -DEDGE_PROFILER_USE_KINETO -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.10.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, 

03/30 21:50:49 - mmengine - INFO - TorchVision: 0.11.0
03/30 21:50:49 - mmengine - INFO - OpenCV: 4.7.0
03/30 21:50:49 - mmengine - INFO - MMEngine: 0.7.0
03/30 21:50:49 - mmengine - INFO - MMCV: 2.0.0rc4
03/30 21:50:49 - mmengine - INFO - MMCV Compiler: GCC 9.3
03/30 21:50:49 - mmengine - INFO - MMCV CUDA Compiler: 11.3
03/30 21:50:49 - mmengine - INFO - MMDeploy: 1.0.0rc3+
03/30 21:50:49 - mmengine - INFO - 

03/30 21:50:49 - mmengine - INFO - **********Backend information**********
03/30 21:50:49 - mmengine - INFO - tensorrt:	8.2.4.2
03/30 21:50:49 - mmengine - INFO - tensorrt custom ops:	Available
03/30 21:50:49 - mmengine - INFO - ONNXRuntime:	None
03/30 21:50:49 - mmengine - INFO - ONNXRuntime-gpu:	1.8.1
03/30 21:50:49 - mmengine - INFO - ONNXRuntime custom ops:	Available
03/30 21:50:49 - mmengine - INFO - pplnn:	None
03/30 21:50:49 - mmengine - INFO - ncnn:	None
03/30 21:50:49 - mmengine - INFO - snpe:	None
03/30 21:50:49 - mmengine - INFO - openvino:	None
03/30 21:50:49 - mmengine - INFO - torchscript:	1.10.0
03/30 21:50:49 - mmengine - INFO - torchscript custom ops:	NotAvailable
03/30 21:50:49 - mmengine - INFO - rknn-toolkit:	None
03/30 21:50:49 - mmengine - INFO - rknn-toolkit2:	None
03/30 21:50:49 - mmengine - INFO - ascend:	None
03/30 21:50:49 - mmengine - INFO - coreml:	None
03/30 21:50:49 - mmengine - INFO - tvm:	None
03/30 21:50:49 - mmengine - INFO - 

03/30 21:50:49 - mmengine - INFO - **********Codebase information**********
03/30 21:50:49 - mmengine - INFO - mmdet:	3.0.0rc6
03/30 21:50:49 - mmengine - INFO - mmseg:	None
03/30 21:50:49 - mmengine - INFO - mmcls:	None
03/30 21:50:49 - mmengine - INFO - mmocr:	None
03/30 21:50:49 - mmengine - INFO - mmedit:	None
03/30 21:50:49 - mmengine - INFO - mmdet3d:	None
03/30 21:50:49 - mmengine - INFO - mmpose:	None
03/30 21:50:49 - mmengine - INFO - mmrotate:	1.0.0rc1
03/30 21:50:49 - mmengine - INFO - mmaction:	None
```

**Comment by dimakan** [comment]
Please note that mAp is getting lower with higher batch. With batch=1 results are good. 
In result with annotations I see drop of detections proportional to batch increase. 
Can it be issue with Rotated batch nms plugin?

**Comment by dimakan** [comment]
FYI: sdk example with batch seems to not do dynamic batch or anything in parallel at all.
Got almost same runtime on sdk with batch 1 and 8.

**Comment by dimakan** [comment]
Any direction or work around?
Please let me know if I can try something

**Comment by dimakan** [comment]
Hi,
I have a new observation. Built a fresh version from main today.
Same issue, but something strange discovered in 3 simple experiments.

**Experiment 1:** 
Duplicated image (image1) and annotation 8 times and run with batch 1 and 8. Results of test.py came same and good!

**Experiment 2:** 
Duplicated different image (image2) and annotation 8 times and run with batch 1 and 8. Results of test.py came same and good!

**Experiment 3:** 
Duplicated  image1 x 4 and  image2 x 4 (with annotations) and run with batch 1 and 8. Results of test.py came for batch 1 came good and batch 8 lost many detection.


**Conclusion**:
it seems that NMS bug causes cross-talk between data in the batch

Please suggest any other steps

Thanks in advance

**Comment by grimoire** [comment]
https://github.com/open-mmlab/mmdeploy/tree/main/csrc/mmdeploy/backend_ops/tensorrt/batched_rotated_nms

**Comment by dimakan** [comment]
> https://github.com/open-mmlab/mmdeploy/tree/main/csrc/mmdeploy/backend_ops/tensorrt/batched_rotated_nms
> https://github.com/open-mmlab/mmdeploy/tree/main/csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms
> Here are all codes about nms. [allClassRotatedNMS.cu](https://github.com/open-mmlab/mmdeploy/blob/main/csrc/mmdeploy/backend_ops/tensorrt/common_impl/nms/allClassRotatedNMS.cu) is modified from allClassNMS.cu.

Any instructions or tips how to run it in debugger would be greatly appreciated. 
Would prefer to dump NMS input and debug it standalone, if possible.

**Comment by grimoire** [comment]
https://github.com/open-mmlab/mmdeploy/blob/2d85be9aa6f90394f3129d4ef1f44b19da47b388/mmdeploy/mmcv/ops/nms_rotated.py#L80

This is the torch-onnx mapping between TensorRT op and torch layer. You can export it to onnx and convert to tensorrt with trtexec. DO NOT use it in torch since it is just a symbolic mapping.

**Comment by dimakan** [comment]
> https://github.com/open-mmlab/mmdeploy/blob/2d85be9aa6f90394f3129d4ef1f44b19da47b388/mmdeploy/mmcv/ops/nms_rotated.py#L80
> 
> This is the torch-onnx mapping between TensorRT op and torch layer. You can export it to onnx and convert to tensorrt with trtexec. DO NOT use it in torch since it is just a symbolic mapping.

Thanks for your help. 
Just to make sure that I understand. 
By “it” you mean the above function or the model?
Still not clear to me how it’ll help me to run it in debugger or dump inputs and outputs of the nms to stdout/file.

**Comment by grimoire** [comment]
This is "how to dump a single node TensorRT model for debugging".  You can create a model and compare it with the groundtruth(without padding):
https://github.com/open-mmlab/mmdeploy/blob/2d85be9aa6f90394f3129d4ef1f44b19da47b388/mmdeploy/mmcv/ops/nms_rotated.py#L9

Sorry I should have to debug it for you, but I am so busy these days. If you have any problem when debugging the op. Please feel free to ask.

**Comment by grimoire** [comment]
> **Experiment 3:** Duplicated image1 x 4 and image2 x 4 (with annotations) and run with batch 1 and 8. Results of test.py came for batch 1 came good and batch 8 lost many detection.

I can reproduce the error with the model but can not reproduce it with a single nms node.

**Comment by grimoire** [comment]
I think I find the solution.
cascade head would update rois multiple time before roi align. The rois is labeled with the batch id (first column).
We do not pad the right batch id here
https://github.com/open-mmlab/mmdeploy/blob/49103cb72e8f9c5ae06159235edc1474fe2c7ce4/mmdeploy/codebase/mmdet/models/roi_heads/cascade_roi_head.py#L72

Please replace it with `rois = torch.cat([batch_index.flatten(0, 1), rois], dim=-1)` and try again.

**Comment by dimakan** [comment]
> I think I find the solution. cascade head would update rois multiple time before roi align. The rois is labeled with the batch id (first column). We do not pad the right batch id here
> 
> https://github.com/open-mmlab/mmdeploy/blob/49103cb72e8f9c5ae06159235edc1474fe2c7ce4/mmdeploy/codebase/mmdet/models/roi_heads/cascade_roi_head.py#L72
> 
> Please replace it with `rois = torch.cat([batch_index.flatten(0, 1), rois], dim=-1)` and try again.

Added this change, run deploy.py again and test.py on the mini test.
Same result for any batch now!
Will do more tests on the full dataset during next couple of days and update ASAP.
Thanks a lot

**Comment by dimakan** [comment]
Full dota test passed!
Please add the fix to code stream
